backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -532,30 +613,89 @@ public FutureExperimentRunDAO(
               return localQueryContext;
             },
             executor);
-
-    final var futurePredicatesContext =
-        predicatesHandler.processPredicates(request.getPredicatesList(), executor);
-
-    final var futureSortingContext =
-        sortingHandler.processSort(request.getSortKey(), request.getAscending());
-
-    final var futureProjectIds =
-        getAllowedProjects(ModelDBActionEnum.ModelDBServiceActions.READ)
-            .thenApply(
-                projIds ->
-                    new QueryFilterContext()
-                        .addCondition("experiment_run.project_id in (<authz_project_ids>)")
-                        .addBind(q -> q.bindList("authz_project_ids", projIds)),
-                executor);
+    queryFilterContextList.add(futureLocalContext);
+
+    // futurePredicatesContext
+    queryFilterContextList.add(
+        predicatesHandler.processPredicates(request.getPredicatesList(), executor));
+
+    // futureSortingContext
+    queryFilterContextList.add(
+        sortingHandler.processSort(request.getSortKey(), request.getAscending()));
+
+    // futureProjectIds based on workspace
+    if (request.getWorkspaceName().isEmpty()) {
+      queryFilterContextList.add(
+          getAllowedProjects(ModelDBActionEnum.ModelDBServiceActions.READ)
+              .thenCompose(
+                  projectIds -> {
+                    if (projectIds.isEmpty()) {
+                      // TODO: This is not works, need to update it.
+                      /*return InternalFuture.failedStage(new PermissionDeniedException(
+                      "Accessible projects not found in workspace: "
+                              + request.getWorkspaceName()));*/
+                    }
+                    return InternalFuture.completedInternalFuture(projectIds);
```

Comment:
@conradoverta , @hmacdonald-verta, I am throwing error from here but it is not works and request is stuck somewhere so i don't understand what i did wrong and where i have writing wrong InternalFuture. Could you check and suggest me here please??

Reply:
I don't unserstand what is not working and how you know it's not working

Reply:
I have try `return InternalFuture.failedStage` but MDB stuck after that statement and if i am trying to throw `throw new PermissionDeniedException` then it is also stuck there. i don't what is happened. Could you suggest me something for it please?